keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,1060 @@
+
+- [shiftfs: uid/gid shifting filesystem](https://lwn.net/Articles/757650/)
+- [A new API for mounting filesystems](https://lwn.net/Articles/753473/)
+- [user_namespace: introduce fsid mappings](https://lwn.net/Articles/812221/)
+
+In regard to this proposal, volumes can be divided in ephemeral and non-ephemeral.
+
+Ephemeral volumes are associated to a **single** pod and their lifecyle is
+dependent on that pod. These are `configMap`, `secret`, `emptyDir`,
+`downwardAPI`, etc. These kind of volumes can work with any of the three
+different modes of `userNamespaceMode` as they are not shared by different pods
+and hence all the processes accessing those volumes have the same effective user
+and group IDs. Kubelet creates the files for those volumes and it can easily set
+the file ownership too.
+
+Non-ephemeral volumes are more difficult to support since they can be persistent
+and shared by multiple pods. This proposal supports volumes with two different
+strategies:
+- The `Cluster` mode makes it easier for pods to share files using volumes when those
+  don't have access permissions for `others` because the effective user and
+  group IDs on the host are the same for all the pods.
+- The semantics of semantics of `fsGroup` are respected, if it's specified it's
+  assumed to be the correct GID in the host and a 1-to-1 mapping entry for the
+  `fsGroup` is added to the GID mappings for the pod.
+
+This KEP doesn't impose any restriction on the different volumes and
+`userNamespaceMode` combinations and leaves it to users to chose the correct
+combinations based on their specific needs.
+
+There are some cases that require special attention from the user. For instance,
+a process inside a pod will not be able to access files with mode `0700` and
+owned by a user different than the effective user of that process in a volume
+that doesn't support the semantics of `fsGroup` (doesn't support
+[`SetVolumeOwnership`](https://github.com/kubernetes/kubernetes/blob/00da04ba23d755d02a78d5021996489ace96aa4d/pkg/volume/volume_linux.go#L42)
+that updates permissions and ownership of the files to be accesible by the
+`fsGroup` group ID). Such pods should be run in `Host` mode.
+
+#### Container Runtime Support
+
+- **Docker**:
+Docker only supports a [single ID
+  mappings](https://docs.docker.com/engine/security/userns-remap/) shared by all
+  containers running in the host. There is not support for [multiple ID
+  mappings](https://github.com/moby/moby/issues/28593) yet. Dockershim runtime
+  is only compatible with pods running in `Host` and `Cluster` modes. The user
+  has to guarantee that the ID mappings configured in Docker through the
+  `userns-remap` parameter and the cluster-wide range configured in the kubelet
+  are the same. The dockershim implementation includes a check to verify that
+  the IDs mapping received from the kubelet are equal to the ones configured in
+  Docker, returning an error otherwise.
```

Comment:
Let's add a statement that making changes to dockershim is out of scope.

Reply:
Done in https://github.com/kinvolk/kubernetes-enhancements/pull/3/commits/2efb472df6145a0398980a48987d5a8f4ab98612